geomesa-convert/geomesa-convert-avro-schema-registry/src/main/scala/org/locationtech/geomesa/convert/avro/registry/AvroSchemaRegistryConverter.scala

```diff
@@ -0,0 +1,138 @@
+
+  // Create schema registry reader from URL string and create Avro reader cache
+  private val schemaRegistryConfig: Option[LoadingCache[Integer, GenericDatumReader[GenericRecord]]] = config.schemaRegistry match {
+    case Some(s) => Some(getReaderCache(new SchemaRegistryReader(s)))
+  }
+
+  // Create Avro reader cache to map schema ID to GenericDatumReader
+  def getReaderCache(schemaRegistry: SchemaRegistryReader): LoadingCache[Integer, GenericDatumReader[GenericRecord]] = {
+    Caffeine
+      .newBuilder()
+      .build(
+        new CacheLoader[Integer, GenericDatumReader[GenericRecord]] {
+          override def load(id: Integer): GenericDatumReader[GenericRecord] = {
+            schemaRegistry.getAvroReaderFromId(id)
+          }
+        }
+      )
+  }
+
+  override protected def parse(is: InputStream, ec: EvaluationContext): CloseableIterator[GenericRecord] = {
+    schemaRegistryConfig match {
+      case Some(s)                  => new GenericRecordSchemaRegistryIterator(is, s, ec.counter)
+    }
+  }
+
+  override protected def values(parsed: CloseableIterator[GenericRecord],
+                                ec: EvaluationContext): CloseableIterator[Array[Any]] = {
+    val array = Array.ofDim[Any](2)
+    parsed.map { record => array(1) = record; array }
+  }
+}
+
+object AvroSchemaRegistryConverter {
+
+  private val MAGIC_BYTE_LENGTH = 1
+  private val SCHEMA_ID_LENGTH = 4
+
+  case class AvroSchemaRegistryConfig(`type`: String,
+                        schemaRegistry: Option[String],
+                        idField: Option[Expression],
+                        caches: Map[String, Config],
+                        userData: Map[String, Expression]) extends ConverterConfig
+
+  sealed trait SchemaConfig
+
+  case class SchemaRegistry(url: String) extends SchemaConfig
+
+  /**
+    * Reads avro records using a cached confluent-style schema registry
+    *
+    * @param is input stream
+    * @param readerCache GenericDatumReader cache
+    * @param counter counter
+    */
+  class GenericRecordSchemaRegistryIterator private [AvroSchemaRegistryConverter] (is: InputStream,
+                                                                     readerCache: LoadingCache[Integer, GenericDatumReader[GenericRecord]],
+                                                                     counter: Counter)
+    extends CloseableIterator[GenericRecord] {
+
+    private val decoder = DecoderFactory.get.binaryDecoder(is, null)
+    private var record: GenericRecord = _
+
+    override def hasNext: Boolean = !decoder.isEnd
+
+    override def next(): GenericRecord = {
+      counter.incLineCount()
+      // Read confluent-style bytes
+      decoder.skipFixed(MAGIC_BYTE_LENGTH)
+      val bytes = new Array[Byte](SCHEMA_ID_LENGTH)
+      decoder.readFixed(bytes, 0, SCHEMA_ID_LENGTH)
+
+      val id = ByteBuffer.wrap(bytes).getInt()
```

Comment:
would it be possible to use the confluent deserialization classes like we do in https://github.com/locationtech/geomesa/blob/master/geomesa-kafka/geomesa-kafka-confluent/src/main/scala/org/locationtech/geomesa/kafka/confluent/ConfluentFeatureSerializer.scala#L55 ? that seems to transparently access the schema registry, instead of rolling our own client.

Reply:
For today, I added a CachedSchemaRegistryClient and also used the KafkaAvroDeserializer to call getById(id). At least now I'm not entirely rolling my own registry client.

I was having trouble getting the kafkaAvroDeserializer.deserialize() working. I can see if I can get this working later.